abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1807 @@
+# Proxy-Wasm ABI v0.2.1 specification
+
+---
+
+Status: **FINAL**
+
+---
+
+# Callbacks and Functions
+
+### Callbacks exposed by the Wasm module
+
+All callbacks (entry points named `proxy_on_<event>`) are optional,
+and are going to be called only if exposed by the Wasm module.
+
+All callbacks, other than the integration functions, include a unique
+context identifier as the first parameter, which can be used to
+distinguish between different contexts.
+
+
+### Functions exposed by the host environment
+
+All functions exposed by the host return [`proxy_status_t`], which
+indicates status of the call (success, invalid memory access, etc.),
+and return values are written into memory pointed by `return_<value>`
+parameters.
+
+
+## Integration
+
+### `proxy_abi_version_0_2_1`
+
+* params:
+  - none
+* returns:
+  - none
+
+Function marker used during linking to advertise support for
+Proxy-Wasm ABI v0.2.1.
+
+This function is never called.
+
+
+### `_initialize`
+
+* params:
+  - none
+* returns:
+  - none
+
+Called when the Wasm module is first loaded.
+
+
+### `main`
+
+* params:
+  - `i32 (uint32_t) ignored`
+  - `i32 (uint32_t) ignored`
+* returns:
+  - `i32 (uint32_t) ignored`
+
+> **Warning**
+> This is called only if [`_initialize`] is also exported.
+
+Called when the Wasm module is first loaded, after [`_initialize`].
+
+
+### `_start`
+
+* params:
+  - none
+* returns:
+  - none
+
+> **Warning**
+> This is called only if [`_initialize`] is not exported.
+
+Called when the Wasm module is first loaded.
+
+
+### `proxy_on_memory_allocate`
+
+* params:
+  - `i32 (size_t) memory_size`
+* returns:
+  - `i32 (uint8_t *) memory_data`
+
+Called to allocate continuous memory buffer of `memory_size` using
+the in-VM memory allocator.
+
+Plugin must return `memory_data` pointing to the start of the allocated
+memory.
+
+Returning `0` indicates failure.
+
+
+### `malloc`
+
+* params:
+  - `i32 (size_t) memory_size`
+* returns:
+  - `i32 (uint8_t *) memory_data`
+
+> **Warning**
+> This is called only in the absence of [`proxy_on_memory_allocate`].
+
+Called to allocate continuous memory buffer of `memory_size` using
+the in-VM memory allocator.
+
+Plugin must return `memory_data` pointing to the start of the allocated
+memory.
+
+Returning `0` indicates failure.
+
+
+## Context lifecycle
+
```

Comment:
needs summary comment that everything in this section are exported by the host, and if the host is expected to export all of them.

Reply:
I've added note that all hostcalls are required.